File: data/src/main/java/com/ahn/data/local/CurrencySelectionDataSource.kt
package com.ahn.data.local

import android.content.Context
import androidx.datastore.core.DataStore
import androidx.datastore.preferences.core.Preferences
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.emptyPreferences
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.ahn.domain.currency.model.CalculatorCurrencySelection
import com.ahn.domain.currency.model.ExchangeCurrencySelection
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map
import kotlinx.io.IOException
import javax.inject.Inject
import javax.inject.Singleton

private val Context.currencySelectionDataStore: DataStore<Preferences> by preferencesDataStore(
    name = "currency_selection"
)

@Singleton
class CurrencySelectionDataSource @Inject constructor(
    @param:ApplicationContext private val context: Context,
) {
    private val calculatorMainCurrencyCodeKey = stringPreferencesKey("calculator_main_currency_code")
    private val calculatorSubCurrencyCodeKey = stringPreferencesKey("calculator_sub_currency_code")
    private val exchangeFromCurrencyCodeKey = stringPreferencesKey("exchange_from_currency_code")
    private val exchangeToCurrencyCodeKey = stringPreferencesKey("exchange_to_currency_code")

    suspend fun getCalculatorSelection(): CalculatorCurrencySelection {
        return context.currencySelectionDataStore.data
            .safePreferences()
            .map { prefs ->
                CalculatorCurrencySelection(
                    mainCode = prefs[calculatorMainCurrencyCodeKey],
                    subCode = prefs[calculatorSubCurrencyCodeKey],
                )
            }
            .first()
    }

    suspend fun saveCalculatorMainCurrencyCode(code: String) {
        context.currencySelectionDataStore.edit { prefs ->
            prefs[calculatorMainCurrencyCodeKey] = code
        }
    }

    suspend fun saveCalculatorSubCurrencyCode(code: String) {
        context.currencySelectionDataStore.edit { prefs ->
            prefs[calculatorSubCurrencyCodeKey] = code
        }
    }

    suspend fun saveCalculatorSelection(mainCode: String, subCode: String) {
        context.currencySelectionDataStore.edit { prefs ->
            prefs[calculatorMainCurrencyCodeKey] = mainCode
            prefs[calculatorSubCurrencyCodeKey] = subCode
        }
    }

    suspend fun getExchangeSelection(): ExchangeCurrencySelection {
        return context.currencySelectionDataStore.data
            .safePreferences()
            .map { prefs ->
                ExchangeCurrencySelection(
                    fromCode = prefs[exchangeFromCurrencyCodeKey],
                    toCode = prefs[exchangeToCurrencyCodeKey],
                )
            }
            .first()
    }

    suspend fun saveExchangeFromCurrencyCode(code: String) {
        context.currencySelectionDataStore.edit { prefs ->
            prefs[exchangeFromCurrencyCodeKey] = code
        }
    }

    suspend fun saveExchangeToCurrencyCode(code: String) {
        context.currencySelectionDataStore.edit { prefs ->
            prefs[exchangeToCurrencyCodeKey] = code
        }
    }

    suspend fun saveExchangeSelection(fromCode: String, toCode: String) {
        context.currencySelectionDataStore.edit { prefs ->
            prefs[exchangeFromCurrencyCodeKey] = fromCode
            prefs[exchangeToCurrencyCodeKey] = toCode
        }
    }

    private fun kotlinx.coroutines.flow.Flow<Preferences>.safePreferences() = catch { exception ->
        if (exception is IOException) {
            emit(emptyPreferences())
        } else {
            throw exception
        }
    }
}

File: ios/Packages/Data/Sources/Data/Currency/Local/CurrencySelectionDataSource.swift
import Domain
import Foundation

/// 화면별 통화 선택을 UserDefaults에 저장하는 데이터소스 (Android `CurrencySelectionDataSource` 대응)
public actor CurrencySelectionDataSource {
    private static let calculatorMainCurrencyCodeKey = "calculator_main_currency_code"
    private static let calculatorSubCurrencyCodeKey = "calculator_sub_currency_code"
    private static let exchangeFromCurrencyCodeKey = "exchange_from_currency_code"
    private static let exchangeToCurrencyCodeKey = "exchange_to_currency_code"

    private let userDefaults: UserDefaults

    public init(userDefaults: UserDefaults = .suite(named: "currency_selection")) {
        self.userDefaults = userDefaults
    }

    public func calculatorSelection() -> CalculatorCurrencySelection {
        CalculatorCurrencySelection(
            mainCode: userDefaults.string(forKey: Self.calculatorMainCurrencyCodeKey),
            subCode: userDefaults.string(forKey: Self.calculatorSubCurrencyCodeKey),
        )
    }

    public func saveCalculatorMainCurrencyCode(_ code: String) {
        userDefaults.set(code, forKey: Self.calculatorMainCurrencyCodeKey)
    }

    public func saveCalculatorSubCurrencyCode(_ code: String) {
        userDefaults.set(code, forKey: Self.calculatorSubCurrencyCodeKey)
    }

    public func saveCalculatorSelection(mainCode: String, subCode: String) {
        userDefaults.set(mainCode, forKey: Self.calculatorMainCurrencyCodeKey)
        userDefaults.set(subCode, forKey: Self.calculatorSubCurrencyCodeKey)
    }

    public func exchangeSelection() -> ExchangeCurrencySelection {
        ExchangeCurrencySelection(
            fromCode: userDefaults.string(forKey: Self.exchangeFromCurrencyCodeKey),
            toCode: userDefaults.string(forKey: Self.exchangeToCurrencyCodeKey),
        )
    }

    public func saveExchangeFromCurrencyCode(_ code: String) {
        userDefaults.set(code, forKey: Self.exchangeFromCurrencyCodeKey)
    }

    public func saveExchangeToCurrencyCode(_ code: String) {
        userDefaults.set(code, forKey: Self.exchangeToCurrencyCodeKey)
    }

    public func saveExchangeSelection(fromCode: String, toCode: String) {
        userDefaults.set(fromCode, forKey: Self.exchangeFromCurrencyCodeKey)
        userDefaults.set(toCode, forKey: Self.exchangeToCurrencyCodeKey)
    }
}
